### DISCUSSION TITLE
extension stopped working on execution of command, throw error: invalid flag dgim

### DISCUSSION THREAD

**Discussion by xkeshav** [microsoft/vscode-discussions#441] [Extension Development QnA]
Developing an extension which use regex heavily and it was working fine when the compelte code written in **activate** method but I want to seggregate the code run the operation on specific command so added below command in package.json  

```json
"contributes": {
    "commands": [
      {
        "command": "css-color-collector.init",
        "category": "color",
        "title": "init collector"
      },
      {
        "command": "css-color-collector.collect",
        "category": "color",
        "title": "collect the colors"
      }
    ],

```

and now when I select that command on the extension window it gives error 

> command '<command name>' resulted in error (invalid flags 'dgim')

check attached screenshot

![image](https://user-images.githubusercontent.com/415219/216802948-ecd88ffb-bb95-43ad-91af-8cdbc158ec24.png)

I have added regex support in **tsconfig.json** and it was working fine 

```json

{
  "compilerOptions": {
    "module": "commonjs",
    "target": "ES2018",
    "outDir": "out",
    "lib": ["ES2022", "es2018.regexp"],
    "sourceMap": true,
    "rootDir": "src",
    "strict": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "noUnusedParameters": true 
  }
}

```

someone helpe me out here.

Thanks.

**Comment by xkeshav** [comment]
the issue is **/d** flag, but my extension use the capture group indices and tht's why /d flag is only way.

**Reply by xkeshav** [reply]
also found that sometime it works and when run the same command twice then it gives error. unable to identify the issue.

**Comment by digitarald** [comment]
Have you tried debugging and pausing on error? That should stop execution right when the error happens.

**Reply by xkeshav** [reply]
Execution stops, in fact it does not start the execution at all, but works after 4 5 trials of running extension for once and again start giving the same error. 

in another stance, I change the flag value then it keeps giving error of old flag value 2-3 times then suddenly works , then I change the flag value in the same debugging session and it works with the new flag once and then stops.


What my guess is that there is something editor not being closed properly or caching of open document file might also causing the issue. 

Thanks. You can check the repo in my profile. *develop* branch

**Reply by xkeshav** [reply]
file: https://github.com/xkeshav/color-collector/blob/develop/src/extension.ts#L53

**Reply by xkeshav** [reply]
Hello,
Any help here?

**Reply by digitarald** [reply]
If debugging fails, have you tried adding logs to see how far it executes?

**Reply by xkeshav** [reply]
I have tried and it fail as soon reach on RegExp.matchAll() method, see attached gif for more understanding.

![Screen Recording 2023-02-15 at 11 15 47 AM](https://user-images.githubusercontent.com/415219/218944166-0b1baf4a-e1db-43ca-adfc-c5831a584057.gif)


the annoying this is that, sometimes it works ( I will share the video soon when it works) and once it stopped working, it keep throwing error for many times.

might be I am somthing missing to close the editor or something else. you can see the code on the repo above mentioned.

Kindly help. I have developed a fully functional and now it stopped working suddenly. :(

**Reply by xkeshav** [reply]
and here is the video when it works fine. what i did here is first i change flag by removing d flag and then run and it works ( surprisingly without d flag) then again add d flag and it works.
there is something  wrong with session. Please check

https://user-images.githubusercontent.com/415219/218956487-704910be-7ae5-4748-9e6e-803b5c2e3cd8.mov

**Reply by xkeshav** [reply]
Kindly check the attched video

**Reply by r3m0t** [reply]
When the debugger is paused on the last breakpoint before the issue, check both "Caught Exceptions" and "Uncaught Exceptions" and then press play. Or, use "Step Into" and "Step Over" instead of pressing play.

**Reply by xkeshav** [reply]
if i check both cauth and unCaught Excpetion then there are lot of error from other extension which are installed in my vs code. to overcome that I wrote "--disable-extensions" in args property of launch.json like below

```js
{
      "name": "Run Extension",
      "type": "extensionHost",
      "request": "launch",
      "args": ["--disable-extensions", "--extensionDevelopmentPath=${workspaceFolder}"],
      "outFiles": ["${workspaceFolder}/out/**/*.js"],
      "preLaunchTask": "${defaultBuildTask}"
    }
```

then still few exception from workebnch/api/node/extensionHostProcess.js and many other file,

see the image
![extension caught exception](https://user-images.githubusercontent.com/415219/219704339-097b2a8d-39a2-40fc-a830-1757e8702147.png)

then I let it pass these error and did F11 and it is working for nowe as I said that sometimes it works perfectly. trying to get the instance when it gives error. will keep updating here.

**Reply by xkeshav** [reply]
Further, I have reverted the  above "--disable-extensions" changes from launch.json and start disabling the extension which giving error or having issue ( with both caught exception or uncaught exception option checked, as you suggested @r3m0t ), and found intresting things is that one extension [Grunttfunggly: Todo Tree extension](https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.todo-tree) is causing this issue, 

I have checked the issue 3 4 times by enabling and disabling the above extension and everytime the "invalid flag issue" occurs whenever this extension is enabled. when this is disabled, it works as expected (so far )

 now unable to understand why an another extension causing issue in other  extension, here is the screenshot and you can see the error trace and how my extension issue transferred to todo tree extension. this is my so far analysis and bit strange too. 

 
![Screenshot_20230217_104248](https://user-images.githubusercontent.com/415219/219720985-1525bb37-f26c-4224-8cd0-278ab8d47735.png)
 and here is  the line from collector.ts which causing issue
 
```js
const { groups: selectorGroup, indices: { groups: selectorIndicesGroup } } = matchingSelector as RegExpMatchArrayWithIndices;
```

while in collector.js file tslint says indices are not in RegexMatchArray while I have used "as" syntax for extented RegexMathArray, is this something wrong here?

Thanks for the help, I will keep updating if I found anything else

**Reply by xkeshav** [reply]
this quetion also report the similar issue with todo tree extension https://stackoverflow.com/a/74327398/155861

**Reply by r3m0t** [reply]
> if i check both cauth and unCaught Excpetion then there are lot of error from other extension which are installed in my vs code.

Don't have it checked from the start. Let the program run to your breakpoint and then check them. That will prevent it from stopping in irrelevant code.

**Comment by Gruntfuggly** [comment]
Sorry my extension is causing you an issue, but it makes no sense to me how this is possible?  Does your extension modify any settings of vscode? The only thing I can think of is that the `files.exclude` setting is being changed which will trigger an update in todo-tree. Could that be happening?

**Reply by xkeshav** [reply]
Thanks for reply but AFAIK my extension update content of css file only and do nothing  except this.  can you please help me how to check whether my extension is changing any vs code settings or not ?
